fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/SocialMediaRepositoryTest.java

```diff
@@ -0,0 +1,128 @@
+
+import io.reactivex.observers.TestObserver;
+
+import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageID;
+import static com.github.dedis.popstellar.testutils.Base64DataUtils.generatePublicKey;
+import static java.util.Collections.*;
+import static org.junit.Assert.*;
+
+public class SocialMediaRepositoryTest {
+
+  private static final String LAO_ID = Lao.generateLaoId(generatePublicKey(), 1000, "LAO");
+
+  private static final Chirp CHIRP_1 =
+      new Chirp(
+          generateMessageID(), generatePublicKey(), "This is a chirp !", 1001, new MessageID(""));
+  private static final Chirp CHIRP_2 =
+      new Chirp(
+          generateMessageID(),
+          generatePublicKey(),
+          "This is another chirp !",
+          1003,
+          new MessageID(""));
+
+  @Test
+  public void addingAChirpAfterSubscriptionUpdatesIds() {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    TestObserver<Set<MessageID>> ids = repo.getChirpsOfLao(LAO_ID).test();
+    // make sure the current last element is an empty set
+    ids.assertValueAt(ids.valueCount() - 1, emptySet());
+    repo.addChirp(LAO_ID, CHIRP_1);
+    // make sure we received a new value : the set containing the chirp
+    ids.assertValueAt(ids.valueCount() - 1, singleton(CHIRP_1.getId()));
+  }
+
+  @Test
+  public void addingChirpBeforeSubscriptionUpdateIds() {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    repo.addChirp(LAO_ID, CHIRP_1);
+    TestObserver<Set<MessageID>> ids = repo.getChirpsOfLao(LAO_ID).test();
+    // The value at subscription contains the first chirp's id
+    ids.assertValueAt(ids.valueCount() - 1, singleton(CHIRP_1.getId()));
+    repo.addChirp(LAO_ID, CHIRP_2);
+    // The value at subscription contains the two chirps' ids
+    ids.assertValueAt(ids.valueCount() - 1, setOf(CHIRP_1.getId(), CHIRP_2.getId()));
+  }
+
+  @Test
+  public void deleteChipDispatchToObservable() throws UnknownChirpException {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    repo.addChirp(LAO_ID, CHIRP_1);
+    TestObserver<Chirp> chirp = repo.getChirp(LAO_ID, CHIRP_1.getId()).test();
+    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
+    assertTrue(repo.deleteChirp(LAO_ID, CHIRP_1.getId()));
+    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1.deleted());
+  }
+
+  @Test
+  public void addChirpWithExistingIdHasNoEffect() throws UnknownChirpException {
+    SocialMediaRepository repo = new SocialMediaRepository();
+    repo.addChirp(LAO_ID, CHIRP_1);
+    TestObserver<Chirp> chirp = repo.getChirp(LAO_ID, CHIRP_1.getId()).test();
+    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
+    Chirp invalidChirp =
+        new Chirp(
+            CHIRP_1.getId(),
+            generatePublicKey(),
+            "This is another chirp !",
+            1003,
+            new MessageID(""));
+    repo.addChirp(LAO_ID, invalidChirp);
+    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
+  }
```

Comment:
here and elsewhere, but even more so for this test, it would help here to clearly separate the `arrange` (pre-condition asserts included) / `act` / `assert` sections of the test
```suggestion
  public void addChirpWithExistingIdHasNoEffect() throws UnknownChirpException {
    // arrange
    SocialMediaRepository repo = new SocialMediaRepository();
    repo.addChirp(LAO_ID, CHIRP_1);
    
    TestObserver<Chirp> chirp = repo.getChirp(LAO_ID, CHIRP_1.getId()).test();
    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
    
    // act
    Chirp invalidChirp =
        new Chirp(
            CHIRP_1.getId(),
            generatePublicKey(),
            "This is another chirp !",
            1003,
            new MessageID(""));
    repo.addChirp(LAO_ID, invalidChirp);
    
    // assert
    chirp.assertValueAt(chirp.valueCount() - 1, CHIRP_1);
  }
```

Reply:
Naive/rhetorical question: is this the case because the chirp is invalid or because we ignore a new ID ? :-)

It might be helpful to document in the repository somewhere that we expect only valid data to be provided at this level and that data validation has already happened before.

Reply:
To me, it can occur if a Chirp message is doubled.

And yes it is expected that the ID is valide at this point.

I will add it is the documentation.